### PR TITLE
Initial V0.1 rBoot Generic

### PR DESCRIPTION
**Pull request by hreintke**
@All : 
This is an alpha implementation of a Generic rBoot layer for easing the use of rBoot and provide a rBoot Bios application to support ESP fully remote.

This PR includes a "RBootClass" which implements :  
- An initial layer around rBoot/rbootHttpUpdate
- An initial way to mount/unmount spiffs filesystems, based on rBoot configuration
- An initial CommandHandler interface to the functionality.
- An example application using the above : rboot_generic

Not much documentation yet, 
- see RBootClass.h for usage & interfaces
- connect with telnet client to esp, rboot commands are then available :
   "rboot info" "rboot load 1 http://10.0.0.201/rom0.bin" "rboot ls" etc
type "rboot" for all options.

All command examples have equivalent API's -> available within application

Current V0.1 is only usable/has hardcoded values for use on 4Mb  devices, no other supported yet.

The used romconfig is the following :

ROM0 -> 0x002000	-> number from initial rboot figures, application area
ROM1 -> 0x202000	-> number from initial rboot figures, application area
ROM2 -> 0x100000;	-> number set in application code, spiffs area, size 65536
ROM3 -> 0x300000;	-> number set in application code, spiffs area, size 65536

Running the example the functionality can be shown by the following commands :

rboot info -> shows ROM info
rboot load 0 http://10.0.0.200/rom0.bin -> error, don't allow current rom update
rboot load 1 http://10.0.0.200/rom0.bin -> http update rom0.bin to ROM1
rboot load 2 http://10.0.0.200/spiffs_rom.bin -> http update spiffs to ROM2
rboot load 3 http://10.0.0.200/spiffs_rom2.bin -> http update spiffs2 to ROM3

rboot start -> start current ROM (= restart)
rboot start 1 -> start ROM1

rboot mount 2 -> mount ROM2 spiffs
rboot ls -> show files on spiffs
rboot unmount -> unmount spiffs
rboot ls -> error not mounted
rboot mount 3 -> mount ROM3 spiffs
rboot ls -> show files on spiffs

Next version to come

Please comment

### CODE REVIEW

**Comment by tprochazka** [comment]
I think that just ESP-12E has 4MB flash or not?

**Comment by hreintke** [comment]
@tprochazka : 
I probably should have left out the (esp-12) note in my initial message.
I took the 4MB flash because of the Basic_Rboot example uses the same.

**Comment by raburton** [comment]
I am interested in this, but just haven't had time to look at anything much Sming related in the last couple of weeks. First I was working nights and then been doing some significant DIY to a deadline.

File: Basic_rBoot _test/.cproject [hunk 1]
```diff
@@ -0,0 +1,151 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (raburton) [inline Basic_rBoot _test/.cproject]: did you mean to include this test project?

File: Basic_rBoot/app/application.cpp [hunk 2]
```diff
@@ -95,7 +95,7 @@ void ShowInfo() {
 void serialCallBack(Stream& stream, char arrivedChar, unsigned short availableCharsCount) {
 	
 
-	if (arrivedChar == '\n') {
+	if (arrivedChar == '\r') {
```
Comment (raburton) [inline Basic_rBoot/app/application.cpp]: why have you made this change? not all systems will send a \r so this is the wrong thing to check for

File: Sming/Makefile-rboot.mk [hunk 3]
```diff
@@ -18,7 +18,7 @@ RBOOT_SPIFFS_1   ?= 0x300000
 RBOOT_LD_0 ?= rom0.ld
 RBOOT_LD_1 ?= rom1.ld
 # esptool2 path
-ESPTOOL2 ?= esptool2
+ESPTOOL2 ?= c:\userdata\esp8266\esptool2\esptool2.exe
```
Comment (raburton) [inline Sming/Makefile-rboot.mk]: You don't want to commit this change. If you want to change this variable for testing without editing the file just set it as an environment variable.

File: rBoot_Generic/readme.txt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+All command examples have equivalent API's -> available within application
+
+Current V0.1 is only usable/has hardcoded values for use on 4Mb (esp-12) devices, no other supported yet.
+
+The used romconfig is the following :
+
+ROM0 -> 0x002000	-> number from initial rboot figures, application area
+ROM1 -> 0x202000	-> number from initial rboot figures, application area
+ROM2 -> 0x100000;	-> number set in application code, spiffs area, size 65536
```
Comment (raburton) [inline rBoot_Generic/readme.txt]: Referring to spiffs files as roms is confusing, I'd avoid that. They are treated differently by rBoot (i.e. rBoot doesn't know about them), and have different purposes. I'd refer to them as spiffs0 & spiffs1.

**Comment by hreintke** [comment]
@raburton : 
Answering to all line comments here.
First : will check my PR's better next time, too many mistakes in this.
- No not Basic_Rboot_test, I noticed that one in the directory and included into my local eclipse, that's why the eclipse files came into the PR -> should not be included
- Maybe better that I rename my example to Basic_rBoot_Generic so it show up next to other rboot examples
- Basic_rboot update : I do have local updates which I use.  -> Should not be included.
- Makefile-rboot -> Same as above
-   Agree on the naming confusion, used as a starting point and did not want to make any changes to rboot-core.
I was thinking of having a "type" included into the romboot_config to distinguish ROM, SPIFFS, OTHER -> would make it into something like a "partition table" 

Questions about PR.
- Should I close this one, and create new when I have next version, 
- Or keep open to get more comments and  push new commits to undo the "unwanted options"

**Comment by raburton** [comment]
Keeping this one is fine, you can commit the fixes then squash the commits to clean it up.

**Comment by alonewolfx2** [comment]
@hreintke what is last status of this pr

**Comment by hreintke** [comment]
@alonewolfx2 
I already did more work on this. Was trying to get a sort of BIOS functionality included but hit some issues which need to be solved. I'll reconsider whether I will first take the RBoot class in.
But need to synchronize with the RTOS activities.

**Comment by hreintke** [comment]
But of course : feel free to take this PR in current state, put in your ideas and add that to the PR

**Comment by hreintke** [comment]
Obsolete, implementing on Sming_RTOS

**Pull request closed without merging**